astropy/table/operations.py

```diff
@@ -810,9 +816,9 @@ def _join(left, right, keys=None, join_type='inner',
             else:
                 # np.where does not work for mixin columns (e.g. Quantity) so
                 # use a slower workaround.
-                left_mask = ~right_mask
-                if np.any(left_mask):
-                    out[out_name][left_mask] = left[left_name].take(left_out)
+                left_mixin_mask = ~right_mask
```

Comment:
This is in fact a bug fix, so I'm happy to resubmit it as a separate PR to be backported. 
(I couldn't really come up with a nice failing examples as I was running into errors (or was using Time columns as key what is a new thing)
```
ValueError: NumPy boolean array indexing assignment cannot assign 6 input values to the 4 output values where the mask is true
```

Reply:
It probably makes sense to split this out so it can be backported, and it would be worthwhile adding a regression test. Here's a minimal test that triggers the error:

```python
from astropy import units as u
from astropy.table import QTable, join

tab1 = QTable()
tab1['index'] = [1, 2, 3, 4, 5]
tab1['flux1'] = [2, 3, 2, 1, 1] * u.Jy
tab1['flux2'] = [2, 3, 2, 1, 1] * u.Jy

tab2 = QTable()
tab2['index'] = [3, 4, 5, 6]
tab2['flux1'] = [2, 1, 1, 3] * u.Jy
tab2['flux2'] = [2, 1, 1, 3] * u.Jy

join(tab1, tab2, keys=('index', 'flux1', 'flux2'), join_type='outer')
```

Reply:
Hmm, this example still doesn't work. But, then it seems that `np.where` now seems to work nicely with Quantities, so no need to go into this branch of the code.